### PR TITLE
feat: add `Functions` field to `options.TerragruntOptions` (#3)

### PR DESCRIPTION
**Pull request by hugorut**
## Description

This PR incorporates a `Functions` field into the `TerragruntOptions` struct. This addition permits Terragrunt library consumers to alter the `hcl.EvaluationContext` functions prior to file evaluation. It proves beneficial for two key use cases:

* Testing: Now, library developers can mock Terragrunt functions in their tests. This avoids sluggish tests and spotlights areas needing improvement in hcl parsing.
* Disable execution: With this new `Functions` feature, consumers can now deactivate function execution in specific contexts. This is essential for us at Infracost, as it allows us to bypass internal Terragrunt and Terraform functions that don't align with our use case and lead to unnecessary program exits. For instance, `get_terraform_cli_args`.

Fixes https://github.com/gruntwork-io/terragrunt/issues/2624

## TODOs

Read the [Gruntwork contribution guidelines](https://gruntwork.notion.site/Gruntwork-Coding-Methodology-02fdcd6e4b004e818553684760bf691e).

- [x] Update the docs. (I don't believe this needs docs updates as it is internal, please let me know if I am wrong)
- [x] Run the relevant tests successfully, including pre-commit checks. (I think so, but when I tried to run the whole test suite locally something was hanging. I don't think these additions will have broken anything, and I've added new tests to cover the logic)
- [x] Ensure any 3rd party code adheres with our [license policy](https://www.notion.so/gruntwork/Gruntwork-licenses-and-open-source-usage-policy-f7dece1f780341c7b69c1763f22b1378) or delete this line if its not applicable.
- [x] Include release notes. If this PR is backward incompatible, include a migration guide.

## Release Notes (draft)

Added new `Functions` field to `TerragruntOptions` to enable `EvalContext` function modification.

### CODE REVIEW

**Comment by hugorut** [comment]
@denis256 any chance of a review on this? It would be great to get some clarity if on if this could be accepted into the core. This just helps us understand if we should commit longer term to maintaining our fork. Thanks

**Comment by hugorut** [comment]
@denis256, just a bump on this. Any clarity you could give would be great. Thanks

File: options/options.go [hunk 1]
```diff
@@ -207,6 +209,15 @@ type TerragruntOptions struct {
...
 	IncludeModulePrefix bool
+
+	// Functions is an optional field which provides a way to add extra functions to the Terragrunt Evaluation Context.
+	// It specifies a function which will be called with the directory of the terragrunt.hcl which is being evaluated.
+	// It should return a map where keys are the names of Terragrunt functions, and values are the corresponding function
+	// implementations. These functions are incorporated into the hcl.EvalContext prior to evaluation.
+	// They always supersede the base functions from Terraform or Terragrunt. This means that the
+	// Functions field can be utilized to overwrite any base functions, which can be particularly
+	// beneficial for testing or cases where certain functions should not be executed.
+	Functions func(baseDir string) map[string]function.Function
```
Comment (denis256) [inline options/options.go:220]: Not sure how much value brings passing only `baseDir` to function

I was thinking about creating a structure that will be passed as inputs with fields like:
 * filename
 * options.TerragruntOptions
 * baseDir

This in future will help to extend fields without affecting the function signature

**Comment by denis256** [comment]
Looks like branch also have conflicts with master

**Comment by hugorut** [comment]
cool thanks @denis256, let me work on the changes in our fork and rebase master. Might take a bit of time as I we'll have to merge a PR into our master to get the changes reflected here.

**Comment by hugorut** [comment]
@denis256 I'm closing this as we've decided to maintain our own fork as we've got some other downstream changes which are only pertinent to our usecase. Thanks for the time spent on this.

**Pull request closed without merging**